### PR TITLE
 Feature(#614):  Challenge38 - Git notes challenge

### LINKED ISSUE #614: Leave a secret in git notes
- [ ] Branch and hide your secret via https://git-scm.com/docs/git-notes
- [ ] Encrypt the secret similar like our other challenges
- [ ] Create the challenge itself explaining how a dev tried to use https://git-scm.com/docs/git-notes to hide a secret.

### PR DESCRIPTION
**Pull request by RemakingEden**
### What kind of changes does this PR include?

-   [ ] Fixes or refactors
-   [x] A new challenge
-   [ ] Additional documentation
-   [ ] Something else

#### Description

This is a challenge to teach users about leaking secrets via git notes.

### Relations

Closes #614 

https://blog.gitguardian.com/leaking-secrets-on-github-what-to-do/

### Checklist:

-   [x] All the contributions made are solely the work of me and my co-authors
-   [x] I tested the changes in this PR (if applicable)
-   [x] I added unit tests to ensure my change works (when change in Java or on front-end code)
-   [x] The PR passes pre-commit hooks and automated tests

### CODE REVIEW

**Comment by RemakingEden** [comment]
@commjoen Weirdly Trufflehog and git secrets isnt picking the secrets in Git notes up. Do you know if this should be possible? It's hard to figure out definitively:

Ive tried:
- Ensure the secret matched a Trufflehog pattern such as an AWS key
- Created a pattern in Git secrets

It seems they are all looking in the committed files instead of the commits themselves.

File: src/main/resources/explanations/challenge34_hint.adoc [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Like other Git challenges this can be solved by manually combing the Git metadata. 
+There are also a plethora of tools that will automatically search for secrets leaked in Git repos.
+
+1. Search manually using `git log`. You can then use a tool such as grep to try and find leaked secrets with regex searches.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge34_hint.adoc:4]: Would it be possible to give the steps here like we do on earlier assignments?
- clone the repo
- ...

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import org.owasp.wrongsecrets.challenges.Spoiler;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+/**
+ * This is a challenge based on leaking secrets with the misuse of Git notes
+ */
+@Component
+@Order(34)
+public class Challenge34 extends Challenge {
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java:18]: Can you rename everything to Challenge35 please? Challenge34 is already in master...

File: src/main/resources/explanations/challenge34_hint.adoc [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+
+2. Search a whole Git repo with a secrets scanning tool such as TruffleHog.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge34_hint.adoc:6]: If the tool does not pick it up, then let's not recommend it. Instead: can we have a look at alternatives? e.g. is there a way to see the notes here in github?

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  public List<RuntimeEnvironment.Environment> supportedRuntimeEnvironments() {
+    return List.of(RuntimeEnvironment.Environment.DOCKER);
+  }
+
+  private String getSolution() {
+    return unobfuscate("UOZFGZTLOLLXHTKEGGS");
+  }
+
+  private String unobfuscate(String obfuscatedString) {
+    final String key = "QWERTYUIOPASDFGHJKLZXCVBNM";
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java:66]: love your plain text shifting!

**Review by commjoen — CHANGES_REQUESTED** [review]
See the comments! Love the setup, but a few things need to change I guess (maybe sync the fork with this one?). I also cannot find teh notes at the commit. did you push all the notes like described in https://stackoverflow.com/questions/18268986/git-how-to-push-messages-added-by-git-notes-to-the-central-git-server ?

File: src/main/resources/explanations/challenge34_hint.adoc (hunk 1 above)
Comment (RemakingEden) [reply src/main/resources/explanations/challenge34_hint.adoc:4]: Sure, i can make them more verbose :) Do you have an example of a good hint doc?

Comment (commjoen) [reply src/main/resources/explanations/challenge34_hint.adoc:4]: https://github.com/OWASP/wrongsecrets/blob/3a479889de50cc23551c2a2a98c565b7560b2253/src/main/resources/explanations/challenge1_hint.adoc#L4

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java (hunk 2 above)
Comment (RemakingEden) [reply src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge34.java]: This is done now :)

**Comment by RemakingEden** [comment]
@commjoen Okay so git notes have now been pushed. Unfortunately it seems they are very hard to work with. It is not possible to see them in Github and they dont automatically get fetched. You need to use the `git fetch origin refs/notes/commits:refs/notes/commits` command

File: src/main/resources/explanations/challenge35.adoc [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<<<<<<< HEAD
+=== Generating Random Keys
+
+Many security folks teach engineers to use secure key derivation functions like https://en.wikipedia.org/wiki/PBKDF2[PBKDF2] when a key needs to be generated. A developer followed this instruction and tried to create a key in `Challenge34.java`, which should now be far more secure than a hardcoded key.
+Can you spot the mistake? Can you find the value of the generated key?
+=======
+=== Git Notes
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35.adoc:7]:
```suggestion

```

File: src/main/resources/explanations/challenge35.adoc [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<<<<<<< HEAD
+=== Generating Random Keys
+
+Many security folks teach engineers to use secure key derivation functions like https://en.wikipedia.org/wiki/PBKDF2[PBKDF2] when a key needs to be generated. A developer followed this instruction and tried to create a key in `Challenge34.java`, which should now be far more secure than a hardcoded key.
+Can you spot the mistake? Can you find the value of the generated key?
+=======
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35.adoc:6]:
```suggestion

```
can you please remove this part?

File: src/main/resources/explanations/challenge35.adoc [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+=== Git Notes
+
+Git commits can be a constant pain point.
+It is fine using a short message, unintelligible garble or mashing the keyboard in a git message until you have the unfortunate task of reviewing past commits and trying to figure out what has happened.
+Git notes is here to solve this, it has been around for a while but often gets overlooked.
+Add extra metadata about the commit without affecting the commit message itself.
+
+Like all Git, once information is commited it is very very hard to remove all reference of it. What could possible go wrong?
+>>>>>>> cefa8809 ( Feature(#614): Added test files)
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35.adoc:15]:
```suggestion
 
```
can you please remove this part?

File: src/main/resources/explanations/challenge35_hint.adoc [hunk 8]
```diff
@@ -0,0 +1,15 @@
+<<<<<<< HEAD
+This challenge can be solved by replaying the Key derivation function with the given inputs.
+
+1. Run the function online
+- Locate the parameters used for the key derivation function in the `generateKey` function in Challenge33.java
+- Copy the used parameters to an online https://www.dcode.fr/pbkdf2-hash[generator] and execute it
+- The website will return the value of the key.
+=======
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35_hint.adoc:8]:
```suggestion
 
```
 looks slike your sync went wrong?

File: src/main/resources/explanations/challenge35_hint.adoc [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+Like other Git challenges this can be solved by manually combing the Git metadata.
+There are also a plethora of tools that will automatically search for secrets leaked in Git repos.
+
+1. Search manually using `git log`. You can then use a tool such as grep to try and find leaked secrets with regex searches.
+
+2. Search a whole Git repo with a secrets scanning tool such as TruffleHog.
+>>>>>>> cefa8809 ( Feature(#614): Added test files)
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35_hint.adoc:15]:
```suggestion
 
```

File: src/main/resources/explanations/challenge35_reason.adoc [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<<<<<<< HEAD
+*Why Key Derivation Functions are not safe when using hardcoded values*
+
+Key Derivation Functions (KDFs) are deterministic. This means that they will always give back the same output for a given input. So, if the parameters are hardcoded, anyone with access to the code can run the KDF with the specified parameters and get the key.
+
+KDFs should be used to generate keys based on dynamic input, such as human-supplied passwords. KDFs might be used when secure random sources can be used to generate keys. Again, the KDF's input is not hardcoded/deterministic in both cases.
+=======
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35_reason.adoc:7]:
```suggestion
 
```

File: src/main/resources/explanations/challenge35_reason.adoc [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+*Why you should be careful with Git notes?*
+
+When developers use Git notes they can also expose secrets if developers inadvertently include sensitive information, such as passwords or API keys, in the notes.
+Unlike regular commits, notes are not easily visible in code reviews, so they can go unnoticed for a long time.
+
+Although manually going through Git notes to look for secrets would be cumbersome, there are many tools attackers could use to search through Git repo's to look for these leaked secrets.
+>>>>>>> cefa8809 ( Feature(#614): Added test files)
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35_reason.adoc:14]:
```suggestion

```

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge35.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+import org.owasp.wrongsecrets.challenges.ChallengeTechnology;
+import org.owasp.wrongsecrets.challenges.Difficulty;
+import org.owasp.wrongsecrets.challenges.Spoiler;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+/** This is a challenge based on leaking secrets with the misuse of Git notes */
+@Component
+@Order(34)
+public class Challenge34 extends Challenge {
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge35.java:16]:
```suggestion
@Order(35)
public class Challenge35 extends Challenge {
```

File: src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge35.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+  public Challenge34(ScoreCard scoreCard) {
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/docker/Challenge35.java:18]:
```suggestion
  public Challenge35(ScoreCard scoreCard) {
```

File: src/main/resources/explanations/challenge35_hint.adoc [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+1. Run the function online
+- Locate the parameters used for the key derivation function in the `generateKey` function in Challenge33.java
+- Copy the used parameters to an online https://www.dcode.fr/pbkdf2-hash[generator] and execute it
+- The website will return the value of the key.
+=======
+Like other Git challenges this can be solved by manually combing the Git metadata.
+There are also a plethora of tools that will automatically search for secrets leaked in Git repos.
+
+1. Search manually using `git log`. You can then use a tool such as grep to try and find leaked secrets with regex searches.
```
Comment (commjoen) [inline src/main/resources/explanations/challenge35_hint.adoc:12]:
```suggestion
1. Search manually using `git log`. You can then use a tool such as grep to try and find leaked secrets with regex searches.
- Clone the repository (`git clone git@github.com:OWASP/wrongsecrets.git`)
- cd into the directory (`cd wrongsecrets`)
- Fetch the notes (execute `git fetch origin refs/notes/commits:refs/notes/commits` and `git fetch origin "refs/notes/*:refs/notes/*"`)
- now do a `git log` and search for the secret (in this case, the keyword is a password: ` git log | grep -A 2 -B 2 password`)
```

**Review by commjoen — CHANGES_REQUESTED** [review]
Hi @RemakingEden , thank you for the updates! can you please add the required unit test after applying the suggestions? See https://github.com/OWASP/wrongsecrets/blob/master/CONTRIBUTING.md#step-3-adding-test-file for more details.

**Comment by RemakingEden** [comment]
This is on hold due to some questions around how heavily Git Notes is used.

**Comment by commjoen** [comment]
Elligible for `hacktoberfest-accepted` label if timely finished in october

**Review by commjoen — APPROVED** [review]
Love this! Thank you!

**Pull request merged**